optimized s_mp_sqr

czurnieden opened the pull request:
`s_mp_sqr` is significantly slower than `s_mp_mul` here (without comba). Changing the addition `r + r` to `r<<1` made it already faster and changing the computation in the carry propagating loop to a manual multi-word doubled the speed-up already gained. The relation between `s_mp_mul` and `s_mp_sqr` follows the theoretical runtimes much closer now.
A test for `s_mp_sqr` against `s_mp_mul` has been added to `etc/test.c`.

Those advantages depend on compiler and/or architectur, of course, but the changes should not make it worse.

![diff_original_vs_new_s_mp_sqr_vs_s_mp_mul](https://user-images.githubusercontent.com/5593270/104083738-5bb6f180-5241-11eb-8286-2ccd57f9c926.png)

sjaeckel reviewed (commented):
Thanks for reminding me of those PR's!

I rebased after merging #502 and fixed the merge conflict ... Does that merge make sense? Both tests `test_s_mp_sqr()` and `test_s_mp_sqr_comba()` do something slightly different, or should there only be one?

czurnieden commented:
> I rebased after merging https://github.com/libtom/libtommath/pull/502 and fixed the merge conflict ... Does that merge make sense?

As far as I can see: yes.

> Both tests `test_s_mp_sqr()` and `test_s_mp_sqr_comba()` do something slightly different, or should there only be one?

The outer loops are different, that is intentional but `test_s_mp_sqr_comba()` only works if Comba is switched off at compile time. Add a round to the CI-tests?

And, no, I don't have the faintest idea anymore why I added that small inner loop in `test_s_mp_sqr_comba()`. The known edge case is the main point, one test at a random point for the unknown edge cases should suffice.

sjaeckel commented:
> Add a round to the CI-tests?
> 
> And, no, I don't have the faintest idea anymore why I added that small inner loop in `test_s_mp_sqr_comba()`. The known edge case is the main point, one test at a random point for the unknown edge cases should suffice.

you wanna update these tests or should we merge as is?

czurnieden commented:
> you wanna update these tests or should we merge as is?

Merge as it is, because I know me and if I start fiddling around I will find this&hellip;and that&hellip;and those&hellip;and at the end a week is gone ;-)
